pmda/density.py

```diff
@@ -0,0 +1,146 @@
+            smax = np.max(coord, axis=0)
+            # Overwrite smin/smax with user defined values
+            smin, smax = _set_user_grid(self._gridcenter, self._xdim,
+                                        self._ydim, self._zdim, smin, smax)
+        else:
+            # Make the box bigger to avoid as much as possible 'outlier'. This
+            # is important if the sites are defined at a high density: in this
+            # case the bulk regions don't have to be close to 1 * n0 but can
+            # be less. It's much more difficult to deal with outliers.  The
+            # ideal solution would use images: implement 'looking across the
+            # periodic boundaries' but that gets complicated when the box
+            # rotates due to RMS fitting.
+            smin = np.min(coord, axis=0) - self._padding
+            smax = np.max(coord, axis=0) + self._padding
+            BINS = fixedwidth_bins(self._delta, smin, smax)
+            arange = np.transpose(np.vstack((BINS['min'], BINS['max'])))
+            bins = BINS['Nbins']
+            # create empty grid with the right dimensions (and get the edges)
+            grid, edges = np.histogramdd(np.zeros((1, 3)), bins=bins,
+            range=arange, normed=False)
+            grid *= 0.0
+        self._grid = grid
+        self._edges = edges
+        self._arange = arange
+        self._bins = bins
+
+    def _single_frame(self, ts, atomgroups):
+        coord = self.current_coordinates(atomgroups[0], self._atomselection,
+                                         self._update_selection)
+        h, edges = np.histogramdd(coord, bins=self._bins, range=self._arange,
+                                  normed=False)
+        return [h, edges]
+
+    def _conclude(self):
+        self._edges = self._results[0, 1]
+        self._grid = self._results[:, 0].sum(axis=0)
+        self._grid /= float(self._n_frames)
+        metadata = self._metadata if self._metadata is not None else {}
+        metadata['psf'] = self._atomgroup.universe.filename
+        metadata['dcd'] = self._trajectory.filename
+        metadata['atomselection'] = self._atomselection
+        metadata['n_frames'] = self._n_frames
+        metadata['totaltime'] = self._atomgroup.universe.trajectory.totaltime
+        metadata['dt'] = self._trajectory.dt
+        metadata['time_unit'] = mda.core.flags['time_unit']
+        parameters = self._parameters if self._parameters is not None else {}
+        parameters['isDensity'] = False  # must override
+        density = Density(grid=self._grid, edges=self._edges,
+                          units={'length': "Angstrom"},
+                          parameters=parameters,
+                          metadata=metadata)
+        density.make_density()
+        self.density = density
+
+    @staticmethod
+    def _reduce(res, result_single_frame):
+        """ 'append' action for a time series"""
+        if isinstance(res, list) and len(res) == 0:
```

Comment:
Have a look at https://github.com/MDAnalysis/pmda/blob/29e98369e76cdb596d688a83a8a6b0add9dada38/pmda/parallel.py#L410-L421 to see why we need the funny `if isinstance(res, list) and len(res) == 0:`. The `run()` method initializes `res = []` and even though we don't need this list, we have to start with it.